src/sciline/data_graph.py

```diff
@@ -0,0 +1,205 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+from __future__ import annotations
+
+import itertools
+from collections.abc import Iterable
+from types import NoneType
+from typing import Any, Callable, Generator, TypeVar, Union, get_args
+
+import cyclebane as cb
+import networkx as nx
+
+from ._provider import ArgSpec, Provider, ToProvider, _bind_free_typevars
+from .handler import ErrorHandler, HandleAsBuildTimeException
+from .typing import Graph, Key
+
+
+def _as_graph(key: Key, value: Any) -> cb.Graph:
+    """Create a cyclebane.Graph with a single value."""
+    graph = nx.DiGraph()
+    graph.add_node(key, value=value)
+    return cb.Graph(graph)
+
+
+def _find_all_typevars(t: type | TypeVar) -> set[TypeVar]:
+    """Returns the set of all TypeVars in a type expression."""
+    if isinstance(t, TypeVar):
+        return {t}
+    return set(itertools.chain(*map(_find_all_typevars, get_args(t))))
+
+
+def _get_typevar_constraints(t: TypeVar) -> set[type]:
+    """Returns the set of constraints of a TypeVar."""
+    return set(t.__constraints__)
+
+
+def _mapping_to_constrained(
+    type_vars: set[TypeVar],
+) -> Generator[dict[TypeVar, type], None, None]:
+    constraints = [_get_typevar_constraints(t) for t in type_vars]
+    if any(len(c) == 0 for c in constraints):
+        raise ValueError('Typevars must have constraints')
+    for combination in itertools.product(*constraints):
+        yield dict(zip(type_vars, combination))
```

Comment:
```suggestion
        yield dict(zip(type_vars, combination, strict=True))
```
To pre-empt an unhappy ruff.